File: leetcode/1-100/50-powx-n/bench/powbench.py
"""Bench mirror of powbench.kara — recursive fast-power, f64-bits sum sink.
CPython. See ../README.md § Benchmarks."""

import struct


def to_bits(x: float) -> int:
    return struct.unpack("<Q", struct.pack("<d", x))[0]


def fast_pow(x: float, n: int) -> float:
    if n == 0:
        return 1.0
    half = fast_pow(x, n // 2)
    return half * half if n % 2 == 0 else half * half * x


def my_pow(x: float, n: int) -> float:
    return 1.0 / fast_pow(x, -n) if n < 0 else fast_pow(x, n)


MASK64 = (1 << 64) - 1


def main() -> None:
    n_iters, k_reps = 400000, 20
    inv2048 = 2.0**-11  # exact
    acc = 0
    for rep in range(k_reps):
        for i in range(n_iters):
            h = ((i + rep * 7919) * 2654435761) & 2047
            x = 1.0 + h * inv2048
            n = ((i + rep) % 129) - 64
            acc = (acc + to_bits(my_pow(x, n))) & MASK64
    print(acc & 0x7FFFFFFFFFFFFFFF)


if __name__ == "__main__":
    main()

File: leetcode/1-100/50-powx-n/bench/powbench.rs
// Bench mirror of powbench.kara — recursive fast-power, f64-bits sum sink.
// rustc -O. See ../README.md § Benchmarks.

fn fast_pow(x: f64, n: i64) -> f64 {
    if n == 0 {
        return 1.0;
    }
    let half = fast_pow(x, n / 2);
    if n % 2 == 0 {
        half * half
    } else {
        half * half * x
    }
}

fn my_pow(x: f64, n: i64) -> f64 {
    if n < 0 {
        1.0 / fast_pow(x, -n)
    } else {
        fast_pow(x, n)
    }
}

fn main() {
    let n_iters: i64 = 400_000;
    let k_reps: i64 = 20;
    let inv2048 = 0.00048828125_f64; // 2^-11, exact
    let mut acc: u64 = 0;
    for rep in 0..k_reps {
        for i in 0..n_iters {
            let h = ((i + rep * 7919) * 2654435761) & 2047;
            let x = 1.0 + (h as f64) * inv2048;
            let n = ((i + rep) % 129) - 64;
            acc = acc.wrapping_add(my_pow(x, n).to_bits());
        }
    }
    println!("{}", acc & 0x7FFF_FFFF_FFFF_FFFF);
}
